taca_ngi_pipeline/utils/nbis_xml_generator.py

```diff
@@ -0,0 +1,304 @@
+            for lane_stat in fc_obj.get("illumina", {}).get('Demultiplex_Stats', {}).get('Barcode_lane_statistics', []):
+                lane_sample = lane_stat.get("Sample", "")
+                if lane_sample.startswith(self.project['project_id']) and lane_sample in self.samples_delivered and lane_sample not in fc_info["samples"]:
+                    prep_id = "A"
+                    #try and get the prep id for given sample and FC combo
+                    if not self.ignore_lib_prep:
+                        try:
+                            sample_preps_fcs = self.sample_prep_fc_map.get(lane_sample)
+                            prep_id_list = [pid for pid, seqruns in sample_preps_fcs.iteritems() if full_run_id in seqruns]
+                            assert len(prep_id_list) == 1
+                            prep_id = prep_id_list[0]
+                        except AssertionError:
+                            self.LOG.warning("Not able to find prep id for sample '{}' - FC '{}'. Found prep seq map is '{}'".format(lane_sample, full_run_id, sample_preps_fcs))
+                            self.LOG.warning("Generated XML file may not be reliable, double check the file manually and run with '--ignore-lib-prep' option if neccesary")
+                    prep_inst_key = "{}_{}".format(prep_id, instrument_key)
+                    if prep_inst_key not in self.sample_aggregated_stat[lane_sample]:
+                        self.sample_aggregated_stat[lane_sample][prep_inst_key] = {'xml_text': fc_info["instrument"], 'runs': []}
+                    self.sample_aggregated_stat[lane_sample][prep_inst_key]['runs'].append(full_run_id)
+                    fc_info["samples"].append(lane_sample)
+            # this should never happen but good to catch
+            if len(fc_info["samples"]) == 0:
+                self.LOG.warn("No sample were found for project {} in fc {}, this is so weird".format(self.project['project_id'], fc_info['run_name']))
+
+
+    def _set_project_design(self):
+        """ Get project library design and protocol details """
+        # This function is not particularly clever, but this is the best
+        # I was able to do with avilable stuff at the time of writing
+        self.project_design = {}
+        # get application type of project
+        proj_app = self.project.get("details", {}).get("application")
+        # get library construction method and parse neccesary information
+        p_input, p_type, p_option, p_category = [''] * 4
+        lib_meth_match = re.search(r'^(.*?),(.*?),(.*?),(.*?)[\[,](.*)$', self.project.get("details", {}).get("library_construction_method", ""))
+        if lib_meth_match:
+            p_input, p_type, p_option, p_category = lib_meth_match.groups()[:4]
+        # if no library kit found use application type or default string
+        if not p_type or "user" in p_type or "house" in p_type:
+            dp_design = proj_app or p_input if p_input.lower() != "library" else "Sample"
```

Comment:
What is dp short for?

Reply:
I've only migrated the code here, so not sure what dp is